### PR TITLE
feat(NavController): adds interfaces for all lifecycle events

### LINKED ISSUE #7512: NavController lifecycle events interfaces
#### Short description of the problem:

_NavController_'s lifecycle events should have interfaces.
#### What behavior are you expecting?

Provide interfaces, which compel us to implement the corresponding methods like `OnInit` in Angular2, in order to benefit from typing and editor tooling.

**Which Ionic Version?** 1.x or 2.x
2.x

### PR DESCRIPTION
**Pull request by manucorporat**
fixes #7512

### CODE REVIEW

File: src/navigation/nav-controller.ts [hunk 1]
```diff
@@ -6,6 +6,95 @@ import { ViewController } from './view-controller';
...
+}
+
+/**
+ * @name OnCanEnter
+ * @description
+ * Lifecycle hook that is called before the view can enter.
+ * This can be used as a sort of "guard" in authenticated views where you need to check permissions before the view can enter.
+ */
+export interface OnCanEnter {
+  ionViewCanEnter(): Promise<boolean>|boolean;
```
Comment (ysgk) [inline src/navigation/nav-controller.ts]: maybe `Promise<void>|boolean;` is correct?

File: src/navigation/nav-controller.ts [hunk 2]
```diff
@@ -6,6 +6,95 @@ import { ViewController } from './view-controller';
...
+}
+
+/**
+ * @name OnCanLeave
+ * @description
+ * Lifecycle hook that is called before the view can leave.
+ * This can be used as a sort of "guard" in authenticated views where you need to check permissions before the view can leave.
+ */
+export interface OnCanLeave {
+  ionViewCanLeave(): Promise<boolean>|boolean;
```
Comment (ysgk) [inline src/navigation/nav-controller.ts]: the save as above.

File: src/navigation/nav-controller.ts (hunk 1 above)
Comment (manucorporat) [inline src/navigation/nav-controller.ts]: you can actually return resolve(false) and the transition is rejected

Comment (manucorporat) [inline src/navigation/nav-controller.ts]: maybe? ionViewCanLeave(): Promise<boolean|void>|boolean;

Comment (ysgk) [inline src/navigation/nav-controller.ts]: I didn't know that. Thanks.

**Comment by kyr0** [comment]
This would be awesome... could this be resolved and merged please?

**Comment by t-ricci-molecle** [comment]
Any update on this?

**Comment by mhartington** [comment]
Closing for now.

**Pull request closed without merging**